grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1152,69 +1152,90 @@ class StockMovementService {
-                            "Item expected quantity to equal ${pickPageItem.requisitionItem.quantity}"
+                            [pickPageItem.requisitionItem.id] as Object[],
+                            "The quantity you selected for item ${pickPageItem.requisitionItem.id} is different to the expected revised quantity. Please review pick."
                     )
                 }
             }
         }
     }
 
     void importPicklistItems(StockMovement stockMovement, List<PickPageItem> pickPageItems, List<Map> picklistItems) {
-        picklistItems.each { params ->
+        Map<String, List> picklistItemEntriesGroupedByRequisitionItem = picklistItems.groupBy { it.id }
+
+        picklistItemEntriesGroupedByRequisitionItem.each { requisitionItemId, picklistsToImport  ->
             // skip rows with errors
-            if (params.hasErrors()) {
+            if (picklistsToImport.any{ it.hasErrors() }) {
                 return
             }
 
             PickPageItem pickPageItem = pickPageItems.find {
-                it.requisitionItem?.id == params.id
+                it.requisitionItem?.id == requisitionItemId
             }
 
-            AvailableItem availableItem = pickPageItem.getAvailableItem(params.binLocation, params.lotNumber)
-
-            RequisitionItem requisitionItem = pickPageItem.requisitionItem?.modificationItem ?: pickPageItem.requisitionItem
+            removeShipmentItemsForModifiedRequisitionItem(pickPageItem.requisitionItem)
 
-            pickPageItem.picklistItems.each {
-                if (it.id) {
-                    it.quantity = 0
+            picklistsToImport.each { params ->
+                AvailableItem availableItem = pickPageItem.getAvailableItem(params.binLocation, params.lotNumber)
+                Picklist picklist = stockMovement.requisition?.picklist
+                // find existing picklist to update
+                PicklistItem picklistItem = picklist.picklistItems.find {
+                    it.inventoryItem?.id == availableItem.inventoryItem?.id &&
+                            it.binLocation?.id == availableItem.binLocation?.id
                 }
+                createOrUpdatePicklistItem(
+                        pickPageItem.requisitionItem,
+                        picklistItem,
+                        availableItem.inventoryItem,
+                        availableItem.binLocation,
+                        params.quantity,
+                        null,
+                        null
+                )
             }
-            pickPageItem.picklistItems.add(new PicklistItem(
-                    requisitionItem: requisitionItem,
-                    inventoryItem: availableItem.inventoryItem,
-                    binLocation: availableItem.binLocation,
-                    quantity: params.quantity,
-                    sortOrder: pickPageItem.sortOrder
-            ))
+            createMissingShipmentItem(pickPageItem.requisitionItem)
```

Comment:
changes implemnted here meant to follo the way for updating picklist like on 
`StockMovementItemApiController > updatePicklist` where we update individual requisition item.

One problem that I have noticed is that previously importing picklist the old way would not update the shipmentItems where we need to additionally do this: `createMissingShipmentItem(requisitionitem)`

Reply:
Oh shit, good catch. That's very important.

Reply:
We probably need to document this somewhere or centralize it i.e. any time we muck with the picklist it automatically updates the shipment. 

> One problem that I have noticed is that previous implementation of importing picklist would not update the shipmentItems where we need to additionally do this: createMissingShipmentItem(requisitionitem)

Additionally, are you saying this is a current bug in production? If you're not sure we should definitely add this as a discussion point to the ticket with the advise that we probably want to investigate this as an OBS data issue i.e. comparing picklist items to shipment items to ensure they include the same quantity. If not, this is going to be a pain to clean up.

Reply:
Most likely yes, since my previous implementation for the pick import was based on the old import I would assume that the same issue occurred before.